File: ch10/AltZerolessRedundantBinaryRandomAccessList.js
function cons(x, ds) {
    dd = [];
    if (ds.length === 0) {
        dd = [[x]];
    } else if (ds[0].length < 3) {
        dd = ds;
        dd[0].unshift(x);
    } else {
        dd = cons([ds[0][1], ds[0][2]], ds.slice(1));
        dd.unshift([x, ds[0][0]]);
    }
    return dd;
}

function uncons(ds) {
    if (ds.length === 1 && ds[0].length === 1) {
        return [ds[0][0],[]];
    } else if (ds[0].length > 1) {
        dd = ds.slice(1);
        dd.unshift(ds[0].slice(1));
        return [ds[0][0], dd];
    } else {
        u = uncons(ds.slice(1));
        u[1].unshift(u[0]);
        return [ds[0][0], u[1]];
    }
}

function head(xs) {return uncons(xs)[0]}

function tail(xs) {return uncons(xs)[1]}

function show(a) {
    if (Array.isArray(a)) {
        return "[" + a.map(show).join(",") + "]";
    } else {
        return  "" + a;
    }
}

list = [];
for (i = 0; i < 16; i++) {
    list = cons(i, list);
    console.log(show(list));
}
for (i = 0; i < 16; i++) {
    list = tail(list);
    console.log(show(list));
}

File: ch10/AltZerolessRedundantBinaryRandomAccessList.hs
module AltBinaryRandomAccessList (BinaryList) where

import Prelude hiding (head, tail, lookup)
import RandomAccessList

data BinaryList a = Nil
  | One a (BinaryList (a, a))
  | Two a a (BinaryList (a, a))
  | Three a a a (BinaryList (a, a)) deriving Show

uncons:: BinaryList a -> (a, BinaryList a)
uncons Nil = error "empty list "
uncons (One x Nil) = (x, Nil)
uncons (Two t t1 ds) = (t, One t1 ds)
uncons (Three t t1 t2 ds) = (t, Two t1 t2 ds)
uncons (One t ds) =  (t, Two t1 t2 ds')
  where ((t1, t2), ds') = uncons ds

instance RandomAccessList BinaryList where

  empty = Nil

  isEmpty Nil = True
  isEmpty _ = False

  cons x Nil = One x Nil
  cons x (One y ps) = Two x y ps
  cons x (Two y z ps) = Three x y z ps
  cons x (Three y z w ps) = Two x y (cons (z, w) ps)

  head xs = fst (uncons xs)

  tail xs = snd (uncons xs)

  lookup i xs = undefined

  update i y xs = undefined
